docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,320 @@
+strategies must implement.
+
+Some items of the security context only need to be allocated a single time.  For instance if the
+cluster defines that each namespace should have a block of UIDs used for allocation to service
+accounts then the allocator may allocate UID 1000 to service account X for the lifetime of the
+service account.  This allocation can be stored as an annotation on the namespace or service account level so
+that it is uneditable.  When a security context is generated on behalf of security context
+constraints it should first check the annotations for pre-allocated values and use them if appropriate.
+
+Some strategies such as a "allow pods to run as any user" will not need to store annotations.
+
+The allocator will use the configured strategies with interfaces shown below to either validate
+the settings fall within the allowed values or create a new value.
+
+```go
+// SELinuxSecurityContextConstraintsStrategy defines the interface for all SELinux constraint strategies.
+type SELinuxSecurityContextConstraintsStrategy interface {
+	// Generate creates the SELinuxOptions based on constraint rules.
+	Generate(pod *api.Pod, container *api.Container)  (*api.SELinuxOptions, error)
+	// Validate ensures that the specified values fall within the range of the strategy.
+	Validate(pod *api.Pod, container *api.Container) fielderrors.ValidationErrorList
+}
+
+// RunAsUserSecurityContextConstraintsStrategy defines the interface for all uid constraint strategies.
+type RunAsUserSecurityContextConstraintsStrategy interface {
+	// Generate creates the uid based on policy rules.
+	Generate(pod *api.Pod, container *api.Container) (*int64, error)
+	// Validate ensures that the specified values fall within the range of the strategy.
+	Validate(pod *api.Pod, container *api.Container) fielderrors.ValidationErrorList
+}
+```
+
+## Escalating Privileges by an Administrator
+
+An administrator may wish to create a resource in a namespace that runs with
+escalated privileges.  This is similar to saying `sudo make me a pod`.  By allowing security context
+constraints to operate on both the requesting user and pod's service account administrators are able to
+create pods in namespaces with elevated privileges based on the administrator's security context
+constraints.
+
+This also allows the system to guard commands being executed in the non-conforming container.  For
+instance, an `exec` command can first check the security context of the pod against the security
+context constraints of the user or the user's ability to reference a service account.
+If it does not validate then it can block users from executing the command.  Since the validation
+will be user aware administrators would still be able to run the commands that are restricted to normal users.
+
+## Interaction with the Kubelet
+
+In some cases interaction with the kubelet is necessary.  An example of this is a cluster
```

Comment:
TODO: update based on today's discussion with @pmorie.  The `inspectImage` command can be used to view what is set in the docker file.  If an SC reaches the kubelet with no RunAsUser but the SCC policy says it must run as a non-root UID then we can inspect the image, set the RunAsUser, and use the SCC validation normally.